pkg/image/oci/registry_provider.go

```diff
@@ -0,0 +1,92 @@
+package oci
+
+import (
+	"crypto/tls"
+	"fmt"
+	"net/http"
+
+	"github.com/anchore/stereoscope/internal/log"
+	"github.com/anchore/stereoscope/pkg/file"
+	"github.com/anchore/stereoscope/pkg/image"
+	"github.com/google/go-containerregistry/pkg/authn"
+	"github.com/google/go-containerregistry/pkg/name"
+	"github.com/google/go-containerregistry/pkg/v1/remote"
+)
+
+// RegistryImageProvider is a image.Provider capable of fetching and representing a container image fetched from a remote registry (described by the OCI distribution spec).
+type RegistryImageProvider struct {
+	imageStr        string
+	tmpDirGen       *file.TempDirGenerator
+	registryOptions *image.RegistryOptions
+}
+
+// NewRegistryImageProvider creates a new provider instance for a specific image that will later be cached to the given directory.
+func NewRegistryImageProvider(imgStr string, tmpDirGen *file.TempDirGenerator, registryOptions *image.RegistryOptions) *RegistryImageProvider {
+	return &RegistryImageProvider{
+		imageStr:        imgStr,
+		tmpDirGen:       tmpDirGen,
+		registryOptions: registryOptions,
+	}
+}
+
+// Provide an image object that represents the cached docker image tar fetched a registry.
+func (p *RegistryImageProvider) Provide() (*image.Image, error) {
+	log.Debugf("pulling image info directly from registry image=%q", p.imageStr)
+
+	imageTempDir, err := p.tmpDirGen.NewTempDir()
+	if err != nil {
+		return nil, err
+	}
+
+	ref, err := name.ParseReference(p.imageStr)
+	if err != nil {
+		return nil, fmt.Errorf("unable to parse registry reference=%q: %+v", p.imageStr, err)
+	}
+
+	img, err := remote.Image(ref, prepareRemoteOptions(ref, p.registryOptions)...)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create image from registry: %+v", err)
+	}
+
+	return image.NewImage(img, imageTempDir), nil
+}
+
+func prepareRemoteOptions(ref name.Reference, registryOptions *image.RegistryOptions) []remote.Option {
+	var opts []remote.Option
+	if registryOptions.InsecureSkipTLSVerify {
+		t := &http.Transport{
+			// nolint: gosec
+			TLSClientConfig: &tls.Config{InsecureSkipVerify: true},
+		}
+		opts = append(opts, remote.WithTransport(t))
+	}
+	registry := ref.Context().RegistryStr()
+	discoveredConfiguredAuth := false
+	for idx, auth := range registryOptions.Auth {
+		if auth.Authority == registry {
```

Comment:
why does it need to go through all the authentication methods to return all valid ones instead of the first one?

Reply:
This code is currently `break`ing when it finds a valid authentication configuration (matching authority and basic vs bearer token selection). But I should at the very least come back through here and add more comments, mentioning some of the necessary behavior (e.g. these authentication-based options are mutually exclusive options).

This function as a whole is meant to prepare all `remote.Options` which goes beyond only authentication options (there's the transport option at the top of this function).

Reply:
I see that this loop breaks, but instead of looping + appending + breaking, why can't it return instead of a break condition and omit appending? 
When you say "Options goes beyond only authentication" it becomes problematic, because this is treating authentication in the same block as a TLS setting. It seems to me that this would be easier to grasp if connection settings are separate from authentication, which would avoid these constructs for handling them all in one loop.

Reply:
re: "options go beyond auth", these options are from the gcr lib and we don't have a lot of control when it comes to this form factor... we still have to prepare a list of `remote.Option`.

What I can do to make this easier to read is break out the auth options into their own function, which should simplify that logic some by returning early as you are suggesting. Just a note, `prepareRemoteOptions` would still be prepping the TLS config / http transport but would also be additionally invoking this new function dedicated to auth and appending any new options to the existing options list.

Reply:
on the latest commit 👍